I've actually encountered the issue again and this time got the logs, am diving into it but essentially the offender id and `pid_to_kill` are different (5644 vs 5646). Seems like this is because firecracker (5646) is a child of screen (5644), not a direct child of the pytest worker. Calling `waitpid` on firecracker (5646) would ouput an Error, and `wait_process_termination` is treating Errors as "already terminated" instead of waiting for actual process termination here.

So nothing related to having leftover zombie processes I think, and either way this retry mechanism will solve the issue.

```
============================================================================ ERRORS ============================================================================
________________________________________________ ERROR at teardown of test_serial_dos[vmlinux-5.10.257-PCI_ON] _________________________________________________
[gw9] linux -- Python 3.12.3 /opt/venv/bin/python
conftest.py:465: in microvm_factory
    uvm_factory.kill()
        binary_dir = PosixPath('/firecracker/build/cargo_target/x86_64-unknown-linux-musl/release')
        custom_cpu_template = None
        custom_cpu_template_path = None
        netns_factory = <bound method netns_factory.<locals>.NetNsFactory.get of <conftest.netns_factory.<locals>.NetNsFactory object at 0x7f754a217fb0>>
        reap_orphans = None
        record_property = <function record_property.<locals>.sub at 0x7f7549368720>
        report     = {'call': <TestReport 'integration_tests/functional/test_serial_io.py::test_serial_dos[vmlinux-5.10.257-PCI_ON]' when='call' outcome='passed'>,
 'setup': <TestReport 'integration_tests/functional/test_serial_io.py::test_serial_dos[vmlinux-5.10.257-PCI_ON]' when='setup' outcome='passed'>}
        request    = <SubRequest 'microvm_factory' for <Function test_serial_dos[vmlinux-5.10.257-PCI_ON]>>
        results_dir = PosixPath('/firecracker/tests/../test_results/test_serial_dos/test_serial_dos[vmlinux-5.10.257-PCI_ON]')
        uvm_factory = <framework.microvm.MicroVMFactory object at 0x7f7549309cd0>
framework/microvm.py:1424: in kill
    vm.kill()
        self       = <framework.microvm.MicroVMFactory object at 0x7f7549309cd0>
        vm         = <Microvm id=38b2e0f6-2200-4527-8752-0dc72cd7ec93>
framework/microvm.py:385: in kill
    assert not offenders, (
           ^^^^^^^^^^^^^
E   AssertionError: Firecracker reported its pid 5646, which was killed, but there still exist processes using the supposedly dead Firecracker's jailer_id: 
E    5646 /firecracker --id 38b2e0f6-2200-4527-8752-0dc72cd7ec93 --start-time-us 2981927107 --start-time-cpu-us 4886 --parent-cpu-time-us 3556 --enable-pci --log-path fc.log --level Debug --metrics-path fc.ndjson
        _          = '7140'
        cmd        = 'grep 38b2e0f6-2200-4527-8752-0dc72cd7ec93'
        connection = <host_tools.network.SSHConnection object at 0x7f754a1194c0>
        might_be_dead = False
        monitor    = <MemoryMonitor(Thread-151, stopped daemon 140141727897280)>
        offenders  = [' 5646 /firecracker --id 38b2e0f6-2200-4527-8752-0dc72cd7ec93 --start-time-us '
 '2981927107 --start-time-cpu-us 4886 --parent-cpu-time-us 3556 --enable-pci '
 '--log-path fc.log --level Debug --metrics-path fc.ndjson']
        pid_to_kill = 5644
        proc       = ' 7140 grep 38b2e0f6-2200-4527-8752-0dc72cd7ec93'
        self       = <Microvm id=38b2e0f6-2200-4527-8752-0dc72cd7ec93>
        stderr     = ''
        stdout     = (' 5646 /firecracker --id 38b2e0f6-2200-4527-8752-0dc72cd7ec93 --start-time-us '
 '2981927107 --start-time-cpu-us 4886 --parent-cpu-time-us 3556 --enable-pci '
 '--log-path fc.log --level Debug --metrics-path fc.ndjson\n'
 ' 7138 /bin/sh -c ps ax -o pid,cmd -ww | grep '
 '38b2e0f6-2200-4527-8752-0dc72cd7ec93\n'
 ' 7140 grep 38b2e0f6-2200-4527-8752-0dc72cd7ec93\n')
------------------------------------------------------------------------- JSON report --------------------------------------------------------------------------
report saved to: ../test_results/test-report.json
===================================================================== slowest 10 durations =====================================================================
102.67s setup    integration_tests/functional/test_instrumented_firecracker.py::test_log_instrument_firecracker_basic_functionality
64.59s call     integration_tests/functional/test_pause_resume.py::test_kvmclock_ctrl[vmlinux-5.10.257-no-acpi-PCI_OFF]
64.44s call     integration_tests/functional/test_pause_resume.py::test_kvmclock_ctrl[vmlinux-6.1.172-PCI_ON]
63.50s call     integration_tests/functional/test_pause_resume.py::test_kvmclock_ctrl[vmlinux-5.10.257-PCI_ON]
63.49s call     integration_tests/functional/test_pause_resume.py::test_kvmclock_ctrl[vmlinux-5.10.257-no-acpi-PCI_ON]
63.49s call     integration_tests/functional/test_pause_resume.py::test_kvmclock_ctrl[vmlinux-5.10.257-PCI_OFF]
62.23s call     integration_tests/functional/test_pause_resume.py::test_kvmclock_ctrl[vmlinux-6.1.172-PCI_OFF]
54.47s call     integration_tests/security/test_sec_audit.py::test_cargo_audit
39.39s call     integration_tests/functional/test_gdb.py::test_gdb_connects[vmlinux-6.1.172]
37.35s call     integration_tests/functional/test_balloon.py::test_reinflate_balloon[vmlinux-6.1.172-PCI_OFF]
=================================================================== short test summary info ====================================================================
ERROR integration_tests/functional/test_serial_io.py::test_serial_dos[vmlinux-5.10.257-PCI_ON] - AssertionError: Firecracker reported its pid 5646, which was killed, but there still exist processes using the supposedly dead Firecracker's jailer_id:
 5646 /firecracker --id 38b2e0f6-2200-4527-8752-0dc72cd7ec93 --start-time-us 2981927107 --start-time-cpu-us 4886 --parent-cpu-time-us 3556 --enable-pci --log-path fc.log --level Debug --metrics-path fc.ndjson
==================================================== 2276 passed, 205 skipped, 1 error in 776.44s (0:12:56) ====================================================
Filesystem     Inodes IUsed IFree IUse% Mounted on
overlay          192M  332K  192M    1% /
tmpfs             48M    16   48M    1% /sys/fs/cgroup
/dev/nvme2n1p1   192M  332K  192M    1% /boot
tmpfs             48M     1   48M    1% /dev/shm
tmpfs             48M   973   48M    1% /srv
Filesystem      Size  Used Avail Use% Mounted on
overlay         384G   61G  324G  16% /
tmpfs           189G     0  189G   0% /sys/fs/cgroup
/dev/nvme2n1p1  384G   61G  324G  16% /boot
tmpfs           189G     0  189G   0% /dev/shm
tmpfs            32G  3.4G   29G  11% /srv
565M	/srv/jailer
565M	/srv/jailer/firecracker
602M	/firecracker/build/fuzzing/debug
615M	/firecracker/build/gdb/debug
615M	/root
669M	/usr/lib
707M	/firecracker/build/cargo_target
773M	/usr/local/rust/toolchains/nightly-2025-11-21-x86_64-unknown-linux-gnu/lib
816M	/firecracker/build/fuzzing/x86_64-unknown-linux-musl
816M	/firecracker/build/fuzzing/x86_64-unknown-linux-musl/debug
842M	/usr/local/rust/toolchains/nightly-2025-11-21-x86_64-unknown-linux-gnu
850M	/firecracker/build/gdb/x86_64-unknown-linux-musl
850M	/firecracker/build/gdb/x86_64-unknown-linux-musl/debug
882M	/usr/local/rust/toolchains/1.96.0-x86_64-unknown-linux-gnu/lib/rustlib
1.2G	/usr/local/rust/toolchains/1.96.0-x86_64-unknown-linux-gnu/lib
1.3G	/usr/local/rust/toolchains/1.96.0-x86_64-unknown-linux-gnu
1.4G	/firecracker/build/artifacts/s3:--spec.ccfc.min-firecracker-ci-20260610-d300fd253667-0/x86_64/debug
1.4G	/firecracker/build/fuzzing
1.4G	/srv/test_artifacts/debug
1.5G	/firecracker/build/gdb
2.2G	/usr/local/rust/toolchains
2.5G	/srv/test_artifacts
2.7G	/firecracker/build/artifacts
2.7G	/firecracker/build/artifacts/s3:--spec.ccfc.min-firecracker-ci-20260610-d300fd253667-0
2.7G	/firecracker/build/artifacts/s3:--spec.ccfc.min-firecracker-ci-20260610-d300fd253667-0/x86_64
2.7G	/usr/local/rust
2.8G	/usr/local
3.4G	/srv
3.8G	/usr
6.8G	/firecracker/build
7.7G	/firecracker
16G	/
[Firecracker devtool 2026-06-16T13:53:07+0000] Finished test run ...
🚨 Error: The command exited with status 1
user command error: exit status 1
```

--- a/tests/framework/microvm.py
+++ b/tests/framework/microvm.py
@@ -27,7 +27,7 @@ from pathlib import Path
 from typing import Optional
 
 import psutil
-from tenacity import Retrying, retry, stop_after_attempt, wait_fixed
+from tenacity import Retrying, retry, stop_after_attempt, stop_after_delay, wait_fixed
 
 import host_tools.cargo_build as build_tools
 import host_tools.network as net_tools
@@ -343,6 +343,11 @@ class Microvm:
                 continue
             try:
                 os.kill(pid_to_kill, signal.SIGKILL)
+                try:
+                    utils.wait_process_termination(pid_to_kill)
+                except TimeoutError:
+                    utils.dump_proc_state(pid_to_kill)
+                    raise
             except ProcessLookupError:
                 pass
             except OSError:
@@ -355,37 +360,32 @@ class Microvm:
                     self._dump_debug_information(msg)
                     raise
 
-        if self._spawned and self.firecracker_pid:
-            try:
-                utils.wait_process_termination(self.firecracker_pid)
-            except TimeoutError:
-                utils.dump_proc_state(self.firecracker_pid)
-                raise
-
         # if microvm was spawned then check if it gets killed
         if self._spawned:
             # The following logic guards us against the case where `firecracker_pid` for some
             # reason is the wrong PID, e.g. this is a regression test for
             # https://github.com/firecracker-microvm/firecracker/pull/4442/commits/d63eb7a65ffaaae0409d15ed55d99ecbd29bc572
+            # Note: we have to retry a bit because /proc might show killed processes for a brief amount of time.
+            for attempt in Retrying(
+                wait=wait_fixed(0.1), stop=stop_after_delay(1.0), reraise=True
+            ):
+                with attempt:
+                    # filter ps results for the jailer's unique id
+                    _, stdout, _ = utils.check_output(
+                        "ps ax --no-headers -o pid,cmd -ww"
+                    )
 
-            # filter ps results for the jailer's unique id
-            _, stdout, stderr = utils.run_cmd(
-                f"ps ax -o pid,cmd -ww | grep {self.jailer.jailer_id}"
-            )
+                    offenders = []
+                    for proc in stdout.splitlines():
+                        _, cmd = proc.lower().split(maxsplit=1)
+                        if self.jailer.jailer_id in cmd and "firecracker" in cmd:
+                            offenders.append(proc)
 
-            assert not stderr, f"error querying processes using `ps`: {stderr}"
-
-            offenders = []
-            for proc in stdout.splitlines():
-                _, cmd = proc.lower().split(maxsplit=1)
-                if "firecracker" in proc and not cmd.startswith("screen"):
-                    offenders.append(proc)
-
-            # make sure firecracker was killed
-            assert not offenders, (
-                f"Firecracker reported its pid {self.firecracker_pid}, which was killed, but there still exist processes using the supposedly dead Firecracker's jailer_id: \n"
-                + "\n".join(offenders)
-            )
+                    # make sure firecracker was killed
+                    assert not offenders, (
+                        f"Firecracker reported its pid {self.firecracker_pid}, which was killed, but there still exist processes using the supposedly dead Firecracker's jailer_id: \n"
+                        + "\n".join(offenders)
+                    )
 
         if self.uffd_handler and self.uffd_handler.is_running():
             self.uffd_handler.kill()

